Add EfficientLoFTR model

sbucaille opened the pull request:
# What does this PR do?

As discussed earlier, here is the implementation of EfficientLoFTR with their official weight. 
This PR is an anticipation of the release of other weights like MatchAnything https://github.com/zju3dv/MatchAnything
Fixes #36354


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?
 @qubvel 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
initial commit

sbucaille commented:
Hi @qubvel , this PR is ready for a review 🤗

qubvel commented:
Hi @sbucaille! Super cool 🔥  I will review this week!

qubvel reviewed (commented):
Hi @sbucaille! Super exicted to see new model PR 🤗 🚀  Very, very nicely done in terms of code quality and `transformers` standards! I just leave some comments to make it a bit better

sbucaille committed:
Apply suggestions from code review

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

sbucaille commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -101,6 +101,7 @@
         ("dpr", "DPRQuestionEncoder"),
         ("dpt", "DPTModel"),
         ("efficientformer", "EfficientFormerModel"),
+        ("efficientloftr", "EfficientLoFTRForKeypointMatching"),
```
For SuperGlue, we solely have SuperGlueForKeypointMatching, but I'll add an `EfficientLoFTRModel` containing only feature extraction modules and change `EfficientLoFTRForKeypointMatching` with the matching layers

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+    def __init__(self, config: EfficientLoFTRConfig, device="cpu") -> None:
+        super().__init__()
+        self.config = config
+        if hasattr(config, "rope_scaling") and config.rope_scaling is not None:
+            self.rope_type = config.rope_scaling.get("rope_type")
+        else:
+            self.rope_type = "2d"
```
Nope you are right, this is because I wanted to stick as much possible to how RotaryEmbeddings class was implemented elsewhere but in our case we shouldn't have this situation

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+        # Core RoPE block
+        inv_freq_expanded = self.inv_freq[None, None, None, :].float().expand(1, 1, 1, -1)
+        # Force float32 (see https://github.com/huggingface/transformers/pull/29285)
+        device_type = x.device.type
+        device_type = device_type if isinstance(device_type, str) and device_type != "mps" else "cpu"
+        with torch.autocast(device_type=device_type, enabled=False):
+            emb = torch.zeros(1, h, w, self.config.hidden_size // 2)
+            emb[:, :, :, 0::2] = i_position_indices * inv_freq_expanded
+            emb[:, :, :, 1::2] = j_position_indices * inv_freq_expanded
+
+        sin = emb.sin()
+        cos = emb.cos()
+
```
Same here, I wanted to stick to what is done elsewhere, also there may be differences in terms of values if it is not float32 (as stated in the commented line) so I implemented it like that. In the original implementation, they didn't bother, should I keep it this way or remove the autocast ?

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+        return hidden_state
+
+
+class EfficientLoFTRRepVGGBlock(nn.Module):
+    """
+    RepVGG architecture block introduced by the work "RepVGG: Making VGG-style ConvNets Great Again".
+    """
+
+    def __init__(self, config: EfficientLoFTRConfig, in_channels: int, out_channels: int, stride: int = 1) -> None:
+        super().__init__()
+        activation = config.activation_function
+        self.conv1 = EfficientLoFTRConvNormLayer(
+            config, in_channels, out_channels, kernel_size=3, stride=stride, padding=1
+        )
+        self.conv2 = EfficientLoFTRConvNormLayer(
+            config, in_channels, out_channels, kernel_size=1, stride=stride, padding=0
+        )
+        self.identity = nn.BatchNorm2d(in_channels) if in_channels == out_channels and stride == 1 else None
+        self.activation = nn.Identity() if activation is None else ACT2FN[activation]
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        if self.identity is not None:
+            identity_out = self.identity(hidden_states)
+        else:
+            identity_out = 0
+        hidden_states = self.conv1(hidden_states) + self.conv2(hidden_states) + identity_out
+        hidden_states = self.activation(hidden_states)
+        return hidden_states
```
I saw that, but there is a BatchNorm added in the block which makes it different. Notice that I used the RTDetrV2ConvNormLayer on the other hand

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+                all_hidden_states = all_hidden_states + stage_outputs[1]
+            outputs.append(hidden_states)
+
+        # Exclude first stage in outputs
+        outputs = outputs[1:]
+        # Last stage outputs are coarse outputs
+        coarse_features = outputs[-1]
+        # Rest is residual features used in EfficientLoFTRFineFusionLayer
+        residual_features = outputs[:-1]
+        return coarse_features, residual_features, all_hidden_states
```
Agreed, we can have `EfficientLoFTRModel` have this module as backbone and return all the features, and have the `EfficientLoFTRForKeypointMatching` picking the appropriate features to perform the matching

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+        similarity = coarse_features_0 @ coarse_features_1.transpose(-1, -2)
+        similarity = similarity / self.config.coarse_matching_temperature
+
+        if self.config.coarse_matching_skip_softmax:
+            confidence = similarity
+        else:
+            confidence = nn.functional.softmax(similarity, 1) * nn.functional.softmax(similarity, 2)
```
Dual softmax is required for training but not necessary for inference (making the inference even more efficient), since we are not providing a trainable model then we can get rid of this softmax computation

sbucaille replied:
When I read the paper again, it mentionned differences in terms of performance, so I think it is worth keeping both options

qubvel replied:
Ok!

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+        """
+        batch_size, _, channels, height, width = coarse_features.shape
+
+        # (batch_size, 2, channels, height, width) -> (batch_size, 2, height * width, channels)
```
I agree, I'll change all mentions of channels where we are dealing with features and not images

Commit:
fix: various typos, typehints, refactors from suggestions

Commit:
fix: fine_matching method

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+        """
+        batch_size, height0, width0, height1, width1 = scores.shape
+
+        scores = scores.reshape(batch_size, height0 * width0, height1 * width1)
```
Alright I replaced `.reshape` by `.view`, makes sense

sbucaille committed:
Added EfficientLoFTRModel and AutoModelForKeypointMatching class

sbucaille commented:
Hi @qubvel, thanks for the review !
I've addressed some of the comments, some of them still need to be addressed. 
For the data-dependent operations, at first, I wanted to do like SuperGlue, which is indeed using masks. The thing here is that we deal with dense matching, meaning that 1/8th of the pixels are attempted to be matched. Depending on the matching threshold, we have more of less matches. If we keep the dense, we end up carrying around quite large and potentially sparse tensors with a lot of zeroes. This may have negative impact on the computations of scores for example. 
This is why I fell back to something similar to the original implementation which is using indices, because you end up with smaller tensors. Since the model's goal is supposed to be "efficient" I thought that would be the best way 😅 
Anyway, let me try to implement it with masks and I'll benchmark the inference time of the models to have an idea of the difference.

qubvel commented:
Okay, sounds good. Let's benchmark both versions. We should benefit from compilation if we can make it without graph breaks, and that benefit might be much greater than operating on smaller tensors. In addition, that should unlock torch.export. In the worst case, we can have two different paths for the eager and compiled versions.

sbucaille committed:
fix: got rid of compilation breaking instructions

sbucaille committed:
docs: added todo for plot

sbucaille committed:
fix: used correct hub repo

sbucaille commented on src/transformers/models/efficientloftr/modular_efficientloftr.py:
```diff
@@ -0,0 +1,1442 @@
+_CHECKPOINT_FOR_DOC_ = "stevenbucaille/efficient_loftr"
+
+
+def create_meshgrid(
```
Not sure what you mean here, what I see in modeling_rt_detr is the use of `torch.meshgrid` but here we build on top of `torch.meshgrid`

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1419 @@
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+    def extract_one_channel_pixel_values(self, pixel_values: torch.FloatTensor) -> torch.FloatTensor:
```
Not sure what to do here, from the commented pull request, this [PR](https://github.com/huggingface/transformers/pull/25767) is linked and still open, I don't know if you have dealt with this internally, if so please let me know

sbucaille committed:
docs: added comments

sbucaille committed:
fix: run modular

sbucaille committed:
doc: added PyTorch badge

sbucaille committed:
fix: model repo typo in config

sbucaille committed:
fix: make modular

sbucaille commented:
Hi @qubvel,
I finished addressing the comments you made.
About graph breaks, I implemented the full dense version that allows compilation and you were right, with compilation we have better performance than the original sparse version. Even without the compilation I managed to have better performance although this comes at a small memory cost.
Here are the results : 
```
dense - compile :
{'average_forward_time (s)': 0.07417428731918335, 'peak_memory_usage (MB)': 590.228515625}

dense :
{'average_forward_time (s)': 0.09448522090911865, 'peak_memory_usage (MB)': 977.041015625}

sparse - compile : impossible

sparse :
{'average_forward_time (s)': 0.10868784666061401, 'peak_memory_usage (MB)': 563.5390625}
```

Not sure why the quality test is not passing, `make style` does not change anything in my repository, is this from main ? 🤔 
About the other test, I'll check in the coming days 🤗

qubvel commented on src/transformers/models/efficientloftr/modular_efficientloftr.py:
```diff
@@ -0,0 +1,1455 @@
+    pass
+
+
+class EfficientLoFTRRepVGGBlock(nn.Module):
+    """
+    RepVGG architecture block introduced by the work "RepVGG: Making VGG-style ConvNets Great Again".
+    """
+
+    def __init__(self, config: EfficientLoFTRConfig, in_channels: int, out_channels: int, stride: int = 1):
+        super().__init__()
+        activation = config.activation_function
+        self.conv1 = EfficientLoFTRConvNormLayer(
+            config, in_channels, out_channels, kernel_size=3, stride=stride, padding=1
+        )
+        self.conv2 = EfficientLoFTRConvNormLayer(
+            config, in_channels, out_channels, kernel_size=1, stride=stride, padding=0
+        )
+        self.identity = nn.BatchNorm2d(in_channels) if in_channels == out_channels and stride == 1 else None
+        self.activation = nn.Identity() if activation is None else ACT2FN[activation]
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        if self.identity is not None:
+            identity_out = self.identity(hidden_states)
+        else:
+            identity_out = 0
+        hidden_states = self.conv1(hidden_states) + self.conv2(hidden_states) + identity_out
+        hidden_states = self.activation(hidden_states)
+        return hidden_states
```
We have similar module for RT-DETR, however without batchnorm (we should probably correct it there to match original implementation btw).

sbucaille replied:
You mean correct RT-DETR modeling file to match the original implementation ?

sbucaille replied:
Ok, I wanted to reuse the `RTDetrV2RepVggBlock` module but with my version does not depend on `hidden_size` variable which is declared in the init, so when I use modular, it is still there after converting to modeling but `make style` fails as this variable is not used. On the other hand, I can't use `del hidden_states` in modular since it is not declared in my init...

qubvel commented on src/transformers/models/efficientloftr/modular_efficientloftr.py:
```diff
@@ -0,0 +1,1455 @@
+        self.conv2 = EfficientLoFTRConvNormLayer(
+            config, in_channels, out_channels, kernel_size=1, stride=stride, padding=0
+        )
+        self.identity = nn.BatchNorm2d(in_channels) if in_channels == out_channels and stride == 1 else None
```
```suggestion
        self.identity = nn.BatchNorm2d(in_channels) if in_channels == out_channels and stride == 1 else nn.Identity()
```

sbucaille replied:
See comment below

qubvel commented on src/transformers/models/efficientloftr/modular_efficientloftr.py:
```diff
@@ -0,0 +1,1455 @@
+        self.activation = nn.Identity() if activation is None else ACT2FN[activation]
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        if self.identity is not None:
+            identity_out = self.identity(hidden_states)
+        else:
+            identity_out = 0
+        hidden_states = self.conv1(hidden_states) + self.conv2(hidden_states) + identity_out
```
```suggestion
        hidden_states = self.conv1(hidden_states) + self.conv2(hidden_states) + self.identity(hidden_states)
```

sbucaille replied:
That's also what I had in mind when I first implemented it, but this is wrong. It is not identity we want to add to it, it is zero, which is different as identity adds the hidden states to the value

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1446 @@
+
+        hidden_size = config.hidden_size
+        aggregation_sizes = config.aggregation_sizes
+        self.q_aggregation = (
+            nn.Conv2d(
+                hidden_size,
+                hidden_size,
+                kernel_size=aggregation_sizes[0],
+                padding=0,
+                stride=aggregation_sizes[0],
+                bias=False,
+                groups=hidden_size,
+            )
+            if aggregation_sizes[0] != 1
+            else nn.Identity()
+        )
+
+        if aggregation_sizes[1] != 1:
+            self.kv_aggregation = torch.nn.MaxPool2d(kernel_size=aggregation_sizes[1], stride=aggregation_sizes[1])
+        else:
+            self.kv_aggregation = nn.Identity()
+
+        self.norm = nn.LayerNorm(hidden_size)
```
A bit sceptic about this code path, we have [4, 4] in config and Identity is never used.

```suggestion
        if aggregation_sizes[0] != 1:
            self.q_aggregation = nn.Conv2d(
                hidden_size,
                hidden_size,
                kernel_size=aggregation_sizes[0],
                padding=0,
                stride=aggregation_sizes[0],
                bias=False,
                groups=hidden_size,
            )
        else:
            self.q_aggregation = nn.Identity()

        if aggregation_sizes[1] != 1:
            self.kv_aggregation = torch.nn.MaxPool2d(kernel_size=aggregation_sizes[1], stride=aggregation_sizes[1])
        else:
            self.kv_aggregation = nn.Identity()

        self.norm = nn.LayerNorm(hidden_size)
```

sbucaille replied:
I get it, it is always hard to find the best tradeoff between using parameter as if other options were possible (which we can't tell apart from new EfficientLoFTR version that may be release in the future, if any) and getting rid of it. Do you think I should get rid of this parameter and just hardcode it ? 🤔

qubvel replied:
Ok, let's keep it, but rename `aggregation_sizes` to `q_aggregation_kernel_size` and `kv_aggregation_kernel_size`

qubvel replied:
And the same for strides

sbucaille replied:
Done !

qubvel commented on docs/source/en/model_doc/efficientloftr.md:
````diff
@@ -0,0 +1,108 @@
+
+From the outputs, you can visualize the matches between the two images using the following code:
+```python
+processor.plot_keypoint_matching(images, outputs)
+# TODO make plot
+```
````
Let's make this function with PIL instead of matplotlib, we should have everything there with ImageDraw module

sbucaille replied:
Just made the plotting method use PIL (the result is less appealing but at least we don't need matplotlib dependency

qubvel commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -1454,6 +1455,7 @@
     ]
 )
 
+MODEL_FOR_KEYPOINT_MATCHING_MAPPING_NAMES = OrderedDict([("efficientloftr", "EfficientLoFTRForKeypointMatching")])
```
Should we add SuperGlue as well?

sbucaille replied:
Indeed, I added it

qubvel reviewed (commented):
Hey @sbucaille! Thanks for iterating. It's really nice that you've been able to make it compile-friendly 🚀 Overall, it looks almost ready. Great work! I left just a few comments below, the main concern is about the post-processing. Here's what I got:

```
Keypoint at coordinate [-5 -5] in the first image matches with keypoint at coordinate [701 599] in the second image with a score of 0.62890625.
Keypoint at coordinate [-5 -5] in the first image matches with keypoint at coordinate [715 599] in the second image with a score of 0.95458984375.
Keypoint at coordinate [629 400] in the first image matches with keypoint at coordinate [727 600] in the second image with a score of 0.83935546875.
Keypoint at coordinate [642 400] in the first image matches with keypoint at coordinate [741 600] in the second image with a score of 0.258056640625.
```

I got lots of coordinates at [-5, -5]. Should we consider masking them?

P.S. to fix style issues please run `python utils/custom_init_isort.py`
P.S.S. Also discussed with Arthur that it's better to switch back to `#Copied from` statement for this PR (instead of modular), because we reuse so few modules

sbucaille committed:
fix: removed mask values from outputs

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1446 @@
+        super().__init__(config)
+
+        self.config = config
+        self.model = EfficientLoFTRModel(config)
```
Ohh so that's what this attribute is for, got it. I changed it

sbucaille committed:
feat: added plot_keypoint_matching to EfficientLoFTRImageProcessor

sbucaille committed:
feat: added SuperGlueForKeypointMatching to AutoModelForKeypointMatching list

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1446 @@
+        # Force float32 (see https://github.com/huggingface/transformers/pull/29285)
+        device_type = x.device.type
+        device_type = device_type if isinstance(device_type, str) and device_type != "mps" else "cpu"
+        with torch.autocast(device_type=device_type, enabled=False):
+            emb = torch.zeros(1, height, width, self.config.hidden_size // 2)
+            emb[:, :, :, 0::2] = i_position_indices * inv_freq_expanded
+            emb[:, :, :, 1::2] = j_position_indices * inv_freq_expanded
+
+        sin = emb.sin()
+        cos = emb.cos()
+
+        sin = sin.repeat_interleave(2, dim=-1)
+        cos = cos.repeat_interleave(2, dim=-1)
```
I'm not sure to understand your comment here. With this class I tried to match as much as possible existing implementations already present in transformers by removing what is not used in the model. Also their RoPE is a bit different as it is not head dependent

sbucaille committed:
fix: reformat

sbucaille committed:
refactor: renamed aggregation_sizes config parameter into q, kv aggregation kernel size and stride

sbucaille committed:
doc: added q, kv aggregation kernel size and stride doc to config

sbucaille commented:
Hi @qubvel ,
Addressed your comments.
I'll convert back the modeling file into using the `#Copied from` mechanism in the coming days.
Also, I'll need to overwrite the `batching_equivalence` test as different keypoint coordinates from batched and single row inputs cause it to fail.
I'll ping you when it is done.

sbucaille committed:
refactor: converted efficientloftr implementation from modular to copied from mechanism

sbucaille committed:
tests: overwrote batching_equivalence for "keypoints" specific tests

sbucaille committed:
fix: changed EfficientLoFTRConfig import in test_modeling_rope_utils

sbucaille commented:
Hi @qubvel,
I converted back the modeling file using `Copied from` statements.
I added the custom `batching_equivalence` test which takes care of checking how many keypoints are different, similarly to what we did in SuperGlue, also added comments about it.
Let me know what you think

sbucaille commented:
Hey @qubvel , small gentle bump here aswell 🤗

qubvel commented on src/transformers/models/efficientloftr/image_processing_efficientloftr.py:
```diff
@@ -0,0 +1,458 @@
+        target_sizes = [target_sizes[i : i + 2] for i in range(0, len(images), 2)]
+
+        for image_pair, pair_output, target_size in zip(image_pairs, keypoint_matching_output, target_sizes):
+            height0, width0 = target_size[0]
+            height1, width1 = target_size[1]
```
can we take it from `image_pair[0].shape[:2]` and `image_pair[1].shape[:2]`?

sbucaille replied:
Indeed, `target_sizes` is redundant

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1437 @@
+        >>> processor = AutoImageProcessor.from_pretrained("stevenbucaille/efficient_loftr")
+        >>> model = AutoModel.from_pretrained("stevenbucaille/efficient_loftr")
+
+        >>> with torch.no_grad():
+        >>>     inputs = processor(images, return_tensors="pt")
+        >>>     outputs = model(**inputs)
```
```suggestion
        >>> inputs = processor(images, return_tensors="pt")
        >>> with torch.no_grad():
        >>>     outputs = model(**inputs)
```

qubvel commented on tests/models/efficientloftr/test_modeling_efficientloftr.py:
```diff
@@ -0,0 +1,457 @@
+    test_pruning = False
+    test_resize_embeddings = False
+    test_head_masking = False
+    has_attentions = True
```
```suggestion
    has_attentions = True
    test_torch_exportable = True
```

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1437 @@
+        matching_keypoints[:, :, :, 0] = matching_keypoints[:, :, :, 0] / width
+        matching_keypoints[:, :, :, 1] = matching_keypoints[:, :, :, 1] / height
+
+        if not return_dict:
+            return (coarse_matched_indices, coarse_matching_scores, matching_keypoints) + model_outputs[1:]
```
Please see PR https://github.com/huggingface/transformers/pull/36838, the idea is to use `@can_return_tuple` decorator and remove `return_dict` entirely from modeling code

qubvel reviewed (commented):
Great work @sbucaille, almost ready IMO 🚀 just a few nits, and let's add tests for image processors (including visualization, at least that it can be run without error).

P.S. getting an error for `tests/models/efficientloftr/test_modeling_efficientloftr.py::EfficientLoFTRModelTest::test_batching_equivalence` locally, but that's OK if it pass in CI

sbucaille committed:
fix: make fix-copies

sbucaille committed:
fix: make style

sbucaille commented:
Hey @qubvel , just a small bump here.
But by rebasing on the latest change I end up with an error when the script checks for the importability of all auto mappings. After looking around and comparing with what I did previously with SuperGlue I can't figure out what I am missing, any idea ? 🤔

sbucaille committed:
fix: update rope function to make meta tests pass

sbucaille committed:
fix: rename plot_keypoint_matching to visualize_output for clarity

sbucaille committed:
refactor: optimize image pair processing by removing redundant target size calculations

sbucaille commented:
Re @qubvel , just realized I forgot a couple of your comments, my bad, these are addressed now. I updated the rope code because some meta device tests were not passing, all good now except the auto import which I still can't figure out 😅

sbucaille committed:
feat: add EfficientLoFTRImageProcessor to image processor mapping

sbucaille committed:
refactor: removed logger and updated attention forward

sbucaille committed:
refactor: added auto_docstring and can_return_tuple decorators

sbucaille committed:
refactor: update type imports

sbucaille committed:
refactor: update type hints from List/Dict to list/dict for consistency

sbucaille committed:
refactor: update MODEL_MAPPING_NAMES and __all__ to include LightGlue and AutoModelForKeypointMatching

sbucaille committed:
fix: change type hint for size parameter in EfficientLoFTRImageProcessor to Optional[dict]

sbucaille commented:
@qubvel bumping eloftr now that lightglue is merged, I fixed all the details and added `auto_docstring` and `can_return_tuple` decorators as well as cleaned the attention forward to follow the recent changes

qubvel committed:
fix typing

qubvel committed:
Merge branch 'main' into pr/sbucaille/36355

qubvel committed:
fix some typing issues

qubvel committed:
nit

qubvel committed:
a few more typehint fixes

qubvel committed:
Remove output_attentions and output_hidden_states from modeling code

qubvel committed:
else -> elif to support efficientloftr

qubvel committed:
nit

qubvel reviewed (commented):
Thanks! Looks super nice, just tests for image processing are missing, but other than that it looks ready to be merged 🔥 

(I pushed some refactoring for `output_*` kwargs and typehints)

sbucaille committed:
tests: added EfficientLoFTR image processor tests

sbucaille committed:
refactor: reorder functions

sbucaille committed:
chore: update copyright year in EfficientLoFTR test file

sbucaille commented:
These are nice QoL changes, I'll keep them in mind for the next PRs.
I added the image processing tests, these are basically the same as SuperGlue's and LightGlue's with small differences to the keypoint matching mocks.
Also I refactored the modeling code to move the functions the closest to where they are called

qubvel committed:
Use default rope

qubvel committed:
Add docs

qubvel committed:
Update visualization method

qubvel committed:
fix doc order

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+    base_model_prefix = "efficientloftr"
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+    _supports_flash_attn_2 = True
```
```suggestion
    _supports_flash_attn = True
```

qubvel committed:
remove 2d rope test

qubvel reviewed (approved):
Ok, I made final corrections on my side (removed custom rope and used the "default" one), looks good now! @sbucaille thanks for iterating 🤗 

A few things left:
 - Please merge PR's on the Hub (rope config update and model card tags)
 - Make a proper model card
 - Decide whether we are going to move the model to some org / community org

qubvel committed:
Update src/transformers/models/efficientloftr/modeling_efficientloftr.py

qubvel committed:
fix docs

qubvel commented on src/transformers/models/efficientloftr/image_processing_efficientloftr.py:
```diff
@@ -0,0 +1,461 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2025 The HuggingFace Team. All rights reserved.
```

qubvel committed:
Update src/transformers/models/efficientloftr/image_processing_efficientloftr.py

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        current_in_channels = 1
+
+        for i in range(num_stages):
+            out_channels = int(config.stage_block_dims[i] * config.stage_hidden_expansion[i])
```
same for the current in channels, it should be possible to pre-determine these instead of passing them around. 
The config.stage_num_blocks can be accessed as well. 

Let's remove this, config should explicit these for each stage!

sbucaille replied:
Refactored it in d41fb6beee8680c2544570ac00df13abebd71d3a

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        encoder_hidden_states = encoder_hidden_states.reshape(batch_size, -1, embed_dim)
+
+        if position_embeddings is not None:
+            position_embeddings = get_positional_embeddings_slice(aggregated_hidden_states, position_embeddings)
```
this looks a bit messy, the layer producing them should already produce what you need

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        return hidden_states
+
+
+class EfficientLoFTRLocalFeatureTransformerLayer(nn.Module):
```
should inherit from gradient checkpointing!

qubvel replied:
fixed in https://github.com/huggingface/transformers/pull/36355/commits/f59af0dfa22454158934b0dfcabd5f2339261bfc

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+    config_class = EfficientLoFTRConfig
+    base_model_prefix = "efficientloftr"
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
```
there is no reason for that to be false!

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+
+        self.post_init()
+
+    def _get_matches_from_scores(self, scores: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
from the look of it this should not be done at the modeling side, it's not related to weights

ArthurZucker replied:
that's a tough choice... I agree with you, does not hurt too much to keep it here, but can we write this somewhere? (AKA write the motivations as to why these are here)

qubvel replied:
For sure, I will add a comment :+1:

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+
+        return matching_indices, matching_scores
+
+    def _coarse_matching(
```
Same here, should be a post processing to the outputs of the model, unless the model is needed afterwards?

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+
+        return keypoints, matching_scores, matched_indices
+
+    def _get_first_stage_fine_matching(
```
same here!

ArthurZucker reviewed (commented):
Very nice! a few small things to tackle here and there and should be good to go! 🚀

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        if self.identity is not None:
+            identity_out = self.identity(hidden_states)
```
Yeah, it also looks confusing to me, but it's how it's named in the paper and in timm, actually.

https://github.com/huggingface/pytorch-image-models/blob/b2034bb6c57fa6b41fda7398140bf21405361df7/timm/models/byobnet.py#L620

sbucaille commented on src/transformers/models/efficientloftr/image_processing_efficientloftr.py:
```diff
@@ -0,0 +1,461 @@
+
+        return results
+
+    def visualize_keypoint_matching(
```
Btw, in LightGlue's PR, `plot_keypoint_matching` using `matplotlib` got approved and merged, shouldn't we do the same here for consistency ? @qubvel

qubvel replied:
I suggest keep this name, add the same for other keypoint-matching image processors, and deprecate `plot_keypoint_matching`

sbucaille replied:
Should we do that in the same PR ?

sbucaille replied:
Ok I'll deal with that once this one is merged

qubvel committed:
update gradient

qubvel commented on docs/source/en/model_doc/efficientloftr.md:
```diff
@@ -0,0 +1,117 @@
+
+
+url_image1 = "https://raw.githubusercontent.com/magicleap/SuperGluePretrainedNetwork/refs/heads/master/assets/phototourism_sample_images/united_states_capitol_98169888_3347710852.jpg"
+image1 = Image.open(requests.get(url_image1, stream=True).raw)
```
As far as I remember, only a few image processors support loading images by url. Idefics and smolvlm + URLs are supported by `appy_chat_template` in VLMs. There was an initiative from Yoni to support this for all image processors, but we have not observed any demand at the moment. I think we should update to this code to use our internal `load_image` function

```suggestion
from transformers.image_utils import load_image

image1 = load_image(url_image1)
```

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        # (batch_size, seq_len, embed_dim) -> (batch_size, embed_dim, h, w) with seq_len = h * w
+        message = message.permute(0, 2, 1)
+        message = message.reshape(batch_size, embed_dim, aggregated_h, aggregated_w)
+        if self.q_aggregation_kernel_size != 1:
```
+1 on the question

sbucaille replied:
Indeed it wasn't used. Fixed in d842f63bc9f6bebd1f298445c9bdea145c09f97f

sbucaille committed:
refactor: removed unused codepath

qubvel committed:
Add motivation to keep postprocessing in modeling code

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        kv_aggregation_kernel_size = config.kv_aggregation_kernel_size
+        kv_aggregation_stride = config.kv_aggregation_stride
+
+        if q_aggregation_kernel_size != 1:
```
Fixed in d842f63bc9f6bebd1f298445c9bdea145c09f97f

sbucaille committed:
refactor: removed unnecessary variable declarations

sbucaille committed:
docs: use load_image from image_utils

qubvel commented on src/transformers/models/efficientloftr/image_processing_efficientloftr.py:
```diff
@@ -0,0 +1,461 @@
+        do_rescale: bool = True,
+        rescale_factor: float = 1 / 255,
+        do_grayscale: bool = True,
+        **kwargs,
```
it's about `**kwargs` we are adding typing like 

```
**kwargs: Unpack[ImageKwargs]
```

for better DX, but that's not the case for `__init__` in image processors yet, honestly, I don't even know why we need `**kwargs` here 😄

qubvel replied:
https://github.com/qubvel/transformers/blob/eb2335dd0a17c4e77b2547e3331716663ca37b91/src/transformers/processing_utils.py#L159

qubvel replied:
Here is an example of usage:
https://github.com/qubvel/transformers/blob/better-config-typing/src/transformers/models/gemma3n/processing_gemma3n.py#L109

But! that's for __call__ / preprocess method and not related to the current `**kwargs`
As far as I remember, these kwargs are for deserializing private attributes, so, no need to add anything here

qubvel commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        self.original_inv_freq = self.inv_freq
+
+    @torch.no_grad()
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
@sbucaille can you please take a look? As far as I see, we are always slicing for 1/4 of the embeddings, so indeed we can do it in advance

sbucaille replied:
Yes, I was thinking about it, but I'm unsure where you want this precomputer. In Mixtral (and others) there is the `position_ids` that is provided in forward methods up until the Model classes, should this variable be pre-computed in the image processor ?

sbucaille committed:
refactor: moved stage in and out channels computation to configuration

sbucaille committed:
refactor: set an intermediate_size parameter to be more explicit

sbucaille committed:
refactor: removed all mentions of attention masks as they are not used

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        batch_size, _, embed_dim, height, width = hidden_states.shape
+
+        hidden_states = hidden_states.reshape(-1, embed_dim, height, width)
+        if attention_mask is not None:
```
Indeed, it is used nowhere actually, I removed all occurences of attention masks in ec4d65ac0a79b4016db8376b0cb131df343651ca

sbucaille commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1345 @@
+        return attn_output, attn_weights
+
+
+class EfficientLoFTRMLP(nn.Module):
```
Refactored it in 09463ad50480d9be37ef2f4ad1df466be1ff202a

sbucaille committed:
refactor: moved position_embeddings to be computed once in the model instead of every layer

qubvel commented on src/transformers/models/efficientloftr/configuration_efficientloftr.py:
```diff
@@ -143,6 +143,11 @@ def __init__(
         self.stage_num_blocks = stage_num_blocks if stage_num_blocks is not None else [1, 2, 4, 14]
         self.stage_hidden_expansion = stage_hidden_expansion if stage_hidden_expansion is not None else [1, 1, 1, 1]
         self.stage_stride = stage_stride if stage_stride is not None else [2, 1, 2, 2]
+        self.stage_out_channels = [
+            self.stage_block_dims[i] * self.stage_hidden_expansion[i] for i in range(len(self.stage_block_dims))
+        ]
```
can we avoid having thee of them?
- stage_out_channels
- stage_hidden_expansion
- stage_block_dims

I mean we can only have exact values of `stage_out_channels` and remove stage_hidden_expansion and stage_block_dims, right?

qubvel replied:
commented: https://github.com/huggingface/transformers/commit/71fb8c2bfbd50ceb624a2cad2817df49e63b35d3#r162215426

qubvel replied:
Thanks!

sbucaille committed:
refactor: removed unnecessary hidden expansion parameter from config

sbucaille committed:
refactor: removed completely hidden expansions

sbucaille committed:
refactor: removed position embeddings slice function

sbucaille committed:
tests: fixed broken tests because of previous commit

qubvel committed:
fix is_grayscale typehint

qubvel committed:
not refactoring

qubvel committed:
not renaming

qubvel committed:
move h/w to embeddings class

qubvel committed:
Precompute embeddings in init

qubvel commented:
run-slow: efficientloftr

qubvel committed:
Merge branch 'main' into add_eloftr

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36355). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sbucaille committed:
fix: replaced cuda device in convert script to accelerate device

sbucaille committed:
fix: replaced stevenbucaille repo to zju-community

qubvel committed:
Merge branch 'main' into add_eloftr

qubvel committed:
Remove accelerator.device from conversion script

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1310 @@
+        for in_channels, out_channels, num_blocks, stride in zip(
+            config.stage_in_channels, config.stage_out_channels, config.stage_num_blocks, config.stage_stride
+        ):
+            stage = EfficientLoFTRRepVGGStage(config, in_channels, out_channels, num_blocks, stride)
```
I don't want to fight too much on this, but you can still pas the config and have just the id of the stage, and the `EfficientLoFTRRepVGGStage` takes config, and stage_idx, to take the stride, inchannel and outchannels from the config. And this is applicable to `EfficientLoFTRRepVGGBlock` as well, given a stride id can just read the config

sbucaille replied:
I see what you mean now, figuring out parameters should be done as much as possible in configuration. I even applied this to the `EfficientLoFTRFineFusionLayer` layer in eb80048e8d2e3d3b34776a6635952c0420e5c630

sbucaille replied:
On it, is the check about arguments of __init__ being used or is it class attributes defined in __init__ that is checked ?

qubvel replied:
It says that these attributes are not used in `*_modeling.py` file, I do not remember exactly how the check is organized

sbucaille replied:
We should be good now

sbucaille replied:
We should be good now

ArthurZucker commented on src/transformers/models/efficientloftr/modeling_efficientloftr.py:
```diff
@@ -0,0 +1,1310 @@
+        )
+        self.out_conv_layers = nn.ModuleList()
+        for i in range(1, len(stage_block_dims)):
+            out_conv = EfficientLoFTROutConvBlock(config, stage_block_dims[i], stage_block_dims[i - 1])
```
same comment again, if the stablock_dims come from the config, why don't you just pass the config and le layer_idx, and the `EfficientLoFTROutConvBlock` just reads the `stage_idx` ?

sbucaille replied:
Refactored in eb80048e8d2e3d3b34776a6635952c0420e5c630

ArthurZucker reviewed (approved):
Thanks for iterating looks better, 2 last nits and you can merge 🤗

sbucaille committed:
refactor: moved parameter computation in configuration instead of figuring it out when instantiating a Module

sbucaille committed:
fix: removed unused attributes in configuration

sbucaille committed:
fix: missing self

sbucaille commented:
The refactoring changed weight parameters, let me fix that

sbucaille committed:
fix: refactoring and tests

sbucaille committed:
fix: make style
